### PR TITLE
Integrate py-solc, v0.5.0+ contract updates

### PR DESCRIPTION
**Pull request by KPrasch**
*note: `nucypher/py-solc` is incompatible with windows.*

### CODE REVIEW

File: nucypher/blockchain/eth/sol/source/contracts/Issuer.sol [hunk 1]
```diff
@@ -158,15 +158,15 @@ contract Issuer is Upgradeable {
...
+        require(address(uint160(delegateGet(_testTarget, "token()"))) == address(token));
+        require(delegateGet(_testTarget, "miningCoefficient()") == miningCoefficient);
+        require(delegateGet(_testTarget, "lockedPeriodsCoefficient()") == lockedPeriodsCoefficient);
         require(uint32(delegateGet(_testTarget, "secondsPerPeriod()")) == secondsPerPeriod);
         require(uint16(delegateGet(_testTarget, "rewardedPeriods()")) == rewardedPeriods);
         require(uint16(delegateGet(_testTarget, "lastMintedPeriod()")) == lastMintedPeriod);
-        require(uint256(delegateGet(_testTarget, "currentSupply1()")) == currentSupply1);
-        require(uint256(delegateGet(_testTarget, "currentSupply2()")) == currentSupply2);
-        require(uint256(delegateGet(_testTarget, "totalSupply()")) == totalSupply);
+        require(delegateGet(_testTarget, "currentSupply1()") == currentSupply1);
```
Comment (jMyles) [inline nucypher/blockchain/eth/sol/source/contracts/Issuer.sol]: Is this happening because you don't need to cast to `uint256` anymore in 0.5?

File: nucypher/blockchain/eth/sol/source/contracts/UserEscrow.sol [hunk 2]
```diff
@@ -108,17 +108,17 @@ contract UserEscrow is Ownable {
...
-        emit ETHWithdrawn(owner, balance);
+        msg.sender.transfer(balance);
+        emit ETHWithdrawn(msg.sender, balance);
     }
 
     /**
     * @dev Fallback function send all requests to the target proxy contract
     **/
-    function () public payable onlyOwner {
+    function () external payable onlyOwner {
```
Comment (jMyles) [inline nucypher/blockchain/eth/sol/source/contracts/UserEscrow.sol]: Why the change from public to external here?

File: nucypher/blockchain/eth/sol/source/contracts/Issuer.sol (hunk 1 above)
Comment (szotov) [reply nucypher/blockchain/eth/sol/source/contracts/Issuer.sol]: casting was restricted in 0.5, because of different padding for different types (bytes and uint). With this changings method `delegateGet` (in `Upgreadable`) have padding like `uint256` and not how `bytes32`, so changed return type and using of this method

File: nucypher/blockchain/eth/sol/source/contracts/proxy/Dispatcher.sol [hunk 3]
```diff
@@ -81,9 +85,9 @@ contract Dispatcher is Upgradeable {
     * @dev Fallback function send all requests to the target contract.
     * If contract not exists then result will be unpredictable (see DELEGATECALL)
     **/
-    function () public payable {
-        assert(target != 0x0);
-        bool callSuccess = target.delegatecall(msg.data);
+    function () external payable {
```
Comment (jMyles) [inline nucypher/blockchain/eth/sol/source/contracts/proxy/Dispatcher.sol]: Here too - why public to external?

File: nucypher/blockchain/eth/sol/source/contracts/UserEscrow.sol (hunk 2 above)
Comment (szotov) [reply nucypher/blockchain/eth/sol/source/contracts/UserEscrow.sol]: because of breaking changes in v0.5.0
https://solidity.readthedocs.io/en/v0.5.0/050-breaking-changes.html

File: nucypher/blockchain/eth/sol/source/contracts/proxy/Dispatcher.sol [hunk 4]
```diff
@@ -81,9 +85,9 @@ contract Dispatcher is Upgradeable {
...
+        assert(target != address(0));
+        (bool callSuccess,) = target.delegatecall(msg.data);
```
Comment (jMyles) [inline nucypher/blockchain/eth/sol/source/contracts/proxy/Dispatcher.sol]: The addition of these parentheses and comma are foreign to me.  What do we have happening here?  `delegatecall` started returning a tuple?  When previously a returned a single object?

File: nucypher/blockchain/eth/sol/source/contracts/proxy/Dispatcher.sol (hunk 3 above)
Comment (szotov) [reply nucypher/blockchain/eth/sol/source/contracts/proxy/Dispatcher.sol]: the same, now fallback function only external

File: nucypher/blockchain/eth/sol/source/contracts/proxy/Dispatcher.sol (hunk 4 above)
Comment (szotov) [reply nucypher/blockchain/eth/sol/source/contracts/proxy/Dispatcher.sol]: yep, you're right
second value `bytes memory data`

Comment (jMyles) [reply nucypher/blockchain/eth/sol/source/contracts/proxy/Dispatcher.sol]: I don't recall every having seen a tuple with an anonymous member to the left of an assignment operator.  Is this typical in solidity 0.5+?

File: nucypher/blockchain/eth/sol/source/contracts/UserEscrow.sol (hunk 2 above)
Comment (jMyles) [reply nucypher/blockchain/eth/sol/source/contracts/UserEscrow.sol]: I still don't get it though - doesn't this change the visibility of this function?

File: nucypher/blockchain/eth/sol/source/contracts/proxy/Dispatcher.sol (hunk 4 above)
Comment (szotov) [reply nucypher/blockchain/eth/sol/source/contracts/proxy/Dispatcher.sol]: It was added in some 0.4.2? version

File: nucypher/blockchain/eth/sol/source/contracts/UserEscrow.sol (hunk 2 above)
Comment (szotov) [reply nucypher/blockchain/eth/sol/source/contracts/UserEscrow.sol]: yes, `external` is more strict than `public`. `public` means that method also can be called from this contract, but `fallback` function can be called only outside of the contract. So for `fallback` function `public` works like `external` and in 0.5+ only `external` is allowed
